src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,222 @@
+                nodes = marketoService.getProgramsByTag(task.getTagType().get(), task.getTagVallue().get());
+                break;
+            case DATE_RANGE:
+                nodes = marketoService.getProgramsByDateRange(task.getEarliestUpdatedAt().get(),
+                                task.getLatestUpdatedAt().get(),
+                                task.getFilterType().orNull(),
+                                task.getFilterValues().orNull());
+            }
+        }
+        else {
+            nodes = marketoService.getPrograms();
+        }
+        return FluentIterable.from(nodes).transform(MarketoUtils.TRANSFORM_OBJECT_TO_JACKSON_SERVICE_RECORD_FUNCTION).iterator();
+    }
+
+    @Override
+    public ConfigDiff buildConfigDiff(PluginTask task, Schema schema, int taskCount, List<TaskReport> taskReports)
+    {
+        ConfigDiff configDiff = super.buildConfigDiff(task, schema, taskCount, taskReports);
+        // set next next earliestUpdatedAt, latestUpdatedAt
+        if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.DATE_RANGE && task.getIncremental()) {
+
+            DateTime earliest = new DateTime(task.getEarliestUpdatedAt().get());
+            DateTime latest = new DateTime(task.getLatestUpdatedAt().get());
+
+            Duration d = new Duration(earliest, latest);
+            DateTime nextEarliestUpdatedAt = latest.plusSeconds(1);
+
+            configDiff.set("earliest_updated_at", nextEarliestUpdatedAt.toString(DATE_FORMATTER));
+            configDiff.set("report_duration", task.getReportDuration().or(d.getMillis()));
+        }
+        return configDiff;
+    }
+
+    @Override
+    public ServiceResponseMapper<? extends ValueLocator> buildServiceResponseMapper(PluginTask task)
+    {
+        JacksonServiceResponseMapper.Builder builder = JacksonServiceResponseMapper.builder();
+        builder.add("id", Types.LONG)
+                .add("name", Types.STRING)
+                .add("sfdcId", Types.STRING)
+                .add("sfdcName", Types.STRING)
+                .add("description", Types.STRING)
+                .add("createdAt", Types.TIMESTAMP, MarketoUtils.MARKETO_DATE_TIME_FORMAT)
+                .add("updatedAt", Types.TIMESTAMP, MarketoUtils.MARKETO_DATE_TIME_FORMAT)
+                .add("startDate", Types.TIMESTAMP, MarketoUtils.MARKETO_DATE_TIME_FORMAT)
+                .add("endDate", Types.TIMESTAMP, MarketoUtils.MARKETO_DATE_TIME_FORMAT)
+                .add("url", Types.STRING)
+                .add("type", Types.STRING)
+                .add("channel", Types.STRING)
+                .add("folder", Types.JSON)
+                .add("status", Types.STRING)
+                .add("costs", Types.JSON)
+                .add("tags", Types.JSON)
+                .add("workspace", Types.STRING);
+        return builder.build();
+    }
+
+    public enum QueryBy {
+        TAG_TYPE,
+        DATE_RANGE;
+
+        @JsonCreator
+        public static QueryBy of(String value)
```

Comment:
It's not needed to check `null`, also inside of `QueryBy.valueOf` has internal check if the item is not one of enum items, exception will be thrown, should be changed to something like:

```
@JsonCreator
        public static QueryBy of(String value)
        {
            return QueryBy.valueOf(value.toUpperCase());
        }
```